beginner_source/knowledge_distillation_tutorial.py

```diff
@@ -0,0 +1,651 @@
+
+            # Calculate the true label loss
+            label_loss = ce_loss(student_logits, labels)
+
+            # Weighted sum of the two losses
+            loss = feature_map_weight * hidden_rep_loss + ce_loss_weight * label_loss
+
+            loss.backward()
+            optimizer.step()
+
+            running_loss += loss.item()
+
+        print(f"Epoch {epoch+1}/{epochs}, Loss: {running_loss / len(train_loader)}")
+
+# Notice how our test function remains the same here with the one we used in our previous case. We only care about the actual outputs because we measure accuracy.
+
+# Initialize a ModifiedLightNNRegressor
+torch.manual_seed(42)
+modified_light_nn_mse = ModifiedLightNNRegressor(num_classes=10)
+
+# We do not have to train the modified deep network from scratch of course, we just load its weights from the trained instance
+modified_nn_deep_reg = ModifiedDeepNNRegressor(num_classes=10)
+modified_nn_deep_reg.load_state_dict(nn_deep.state_dict())
+
+# Train and test once again
+train_mse_loss(teacher=modified_nn_deep_reg, student=modified_light_nn_mse, train_loader=train_loader, epochs=10, learning_rate=0.001, feature_map_weight=0.25, ce_loss_weight=0.75, device=device)
+test_accuracy_light_ce_and_mse_loss = test_multiple_outputs(modified_light_nn_mse, test_loader, device)
+
+######################################################################
+# It is expected for the final method to work better than CosineLoss because now, we allowed a trainable layer between the teacher and the student, which gives the student some wiggle room when it comes to learning, rather than pushing the student to copy the teacher's representation.
+# Including the extra network is the idea behind hint based distillation.
+
+print(f"Teacher accuracy: {test_accuracy_deep:.2f}%")
+print(f"Student accuracy without teacher: {test_accuracy_light_ce:.2f}%")
+print(f"Student accuracy with CE + KD: {test_accuracy_light_ce_and_kd:.2f}%")
+print(f"Student accuracy with CE + CosineLoss: {test_accuracy_light_ce_and_cosine_loss:.2f}%")
+print(f"Student accuracy with CE + RegressorMSE: {test_accuracy_light_ce_and_mse_loss:.2f}%")
+
+######################################################################
+# None of the methods above increases the number of parameters for the network or inference time, so the performance increase comes at the little cost of calculating gradients during training. In ML applications we mostly care about inference time because training happens before the model deployment.
+# If our lightweight model is still too heavy for deployment we can apply different ideas such as post training quantization. Additional losses can be applied in many tasks, not just classification and one can experiment with quantities like coefficients, temperature or number of neurons.
+# Feel free to tune any numbers in the tutorial above, but keep in mind, if you change the number of neurons / filters chances are a shape mismatch might occur.
+#
```

Comment:
Please a dd a Conclusion section that will list everything the user learned in this tutorial and give pointers to more information.